src/MergeContext.js

```diff
@@ -4,55 +4,73 @@ const Log = require('./Logger.js');
 const GH = require('./GitHubUtil.js');
 const Util = require('./Util.js');
 
-// A result produced by MergeInitiator and MergeFinalizer classes
-class StepResult
+
+// Process() outcome
+class ProcessResult
 {
-    // treat as private; use static methods below instead
-    constructor(delayMs) {
-        assert(delayMs !== undefined);
-        this._delayMs = delayMs;
+    constructor() {
+        this._delayMs = null; // reprocess in that many milliseconds
+        this._prStaged = null; // this PR holds the lock on the staging branch
     }
 
-    // the step is successfully finished
-    static Succeed() {
-        return new StepResult(0);
+    delayed() {
+        return this._delayMs !== null;
     }
 
-    // the step is postponed (and will be resumed some time later)
-    static Suspend() {
-        return new StepResult(-1);
+    delayMs() {
+        assert(this.delayed());
+        return this._delayMs;
     }
 
-    // the step is finished with failure
-    static Fail() {
-        return new StepResult(null);
+    setDelayMsIfAny(msOrNull) {
+        assert(this._delayMs === null);
+        if (msOrNull !== null) {
+            const ms = msOrNull;
+            assert(ms > 0);
+            this._delayMs = ms;
+        }
     }
 
-    // the step is postponed (and will be resumed in delayMs)
-    static Delay(delayMs) {
-        assert(delayMs !== null && delayMs > 0);
-        return new StepResult(delayMs);
+    setPrStaged(bool) {
+        assert(this._prStaged === null);
+        this._prStaged = bool;
     }
 
-    succeeded() {
-        return this._delayMs !== null && this._delayMs === 0;
+    prStaged() {
+        assert(this._prStaged !== null);
+        return this._prStaged;
     }
+}
 
-    failed() {
-        return this._delayMs === null;
-    }
+// Relaying process() outcome to the caller via doProcess() exceptions:
+//
+// Exception          Abandon  Push-Labels  Result-of-process()
+// _exLostControl()     yes      no           approval delay (if any)
+// _exObviousFailure()  yes      yes          approval delay (if any)
+// _exLabeledFailure()  yes      yes          approval delay (if any)
+// _exSuspend()         no       yes          approval delay (if any)
+// _exRetry()           yes      yes          retry or approval delay
```

Comment:
Instantly retried PRs are not really abandoned and probably should not flip/flop labels. I suggest:

```suggestion
// _exRetryNow()        no       no           zero delay
// _exRetrySoon()       yes      yes          retry or approval delay
```

And adjust the rest of the code to match the above. For example, _exRetrySoon() will have a hard-coded delay calculation (instead of doing that in the thrower).